src/main/java/duke/parser/Parser.java

```diff
@@ -0,0 +1,232 @@
+import duke.exceptions.MissingArgumentException;
+import duke.storage.Deadline;
+import duke.storage.Event;
+import duke.storage.Todo;
+
+/**
+ * The Parser class handles the making sense of user commands
+ *
+ * @author Ryan NgWH
+ */
+public class Parser {
+    /**
+     * Converts user input date (in format 'YYYY/MM/DD hh:mm') to Instant
+     *
+     * @param date Date to be converted (in format 'YYYY/MM/DD')
+     * @param time Time to be converted (in format 'hh:mm')
+     *
+     * @return Instant of the specified datetime
+     */
+    public static Instant userDateToInstant(String date, String time)
+            throws NumberFormatException, StringIndexOutOfBoundsException, DateTimeException {
+        return LocalDateTime.of(
+                Integer.parseInt(date.substring(0, 4)),
+                Integer.parseInt(date.substring(5, 7)),
+                Integer.parseInt(date.substring(8, 10)),
+                Integer.parseInt(time.substring(0, 2)),
+                Integer.parseInt(time.substring(3, 5)))
+                .toInstant(OffsetDateTime.now().getOffset());
+    }
+
+    /**
+     * Parse a string and return a command if valid
+     *
+     * @param input String to be parsed
+     *
+     * @return A valid command for the application
+     */
+    public static Command parse(String input) throws DukeException {
+        // Remove whitespaces from input
+        input = input.strip();
+
+        // Split input
+        ArrayList<String> splitInput = new ArrayList<String>(Arrays.asList(input.split(" ")));
+
+        // Remove unncessary whitespace
+        splitInput.removeIf(item -> item == null || "".equals(item));
+
+        String description;
+        List<String> arguments;
+
+        switch (splitInput.get(0).toLowerCase()) {
+        case "bye": // Exit
+            return new ExitCommand();
+
+        case "list": // List tasks
+            // Check if date filter exists
+            if (splitInput.size() > 1) { // List filtered tasks
+                switch (splitInput.get(1)) {
+                case "/date": // Filter by date
+                    try {
+                        Instant filterDate = userDateToInstant(splitInput.get(2), "00:00");
+
+                        return new ListCommand(filterDate);
+                    } catch (NumberFormatException | DateTimeException | IndexOutOfBoundsException e) {
+                        throw new InvalidArgumentException(
+                                "Date/time format is invalid. Please enter the date/time in the format 'YYYY/MM/DD'");
+                    }
+
+                default: // Invalid filter
+                    throw new InvalidArgumentException(
+                            String.format("Unknown argument '%s' for the 'list' command", splitInput.get(1)));
+                }
+            } else { // Return full list
+                return new ListCommand();
+            }
+
+        case "mark": // Mark task
+            if (splitInput.size() <= 1) {
```

Comment:
There should be some better logic for handling size <= 1 instead of repeating it in majority of the cases. Maybe another nested loop? It would improve readability.

Reply:
Will look into it. Thank you for the suggestion!